client/src/telethon/_impl/mtsender/sender.py

```diff
@@ -266,10 +281,46 @@ async def _do_read(self) -> None:
         else:
             self._on_net_read(recv_data)
         finally:
-            self._try_timeout_ping()
+            self._try_ping_timeout()
             self._step_done.set()
 
-    async def _do_write(self) -> None:
+    async def _do_send(self) -> None:
+        self._try_fill_write()
+
+        if self._write_drain_pending:
+            await self._writer.drain()
+            self._on_net_write()
+
+    async def _try_connect(self):
+        attempts = 0
+
+        ip, port = self.addr.split(":")
+
+        while True:
+            try:
+                self._reader, self._writer = await self._connector(ip, int(port))
+                self._logger.info(
+                    f"auto-reconnect success after {attempts} failed attempt(s)"
+                )
+                return
+            except Exception as e:
+                attempts += 1
+                self._logger.warning(f"auto-reconnect failed {attempts} time(s): {e!r}")
+                await asyncio.sleep(1)
+
+                delay = self._reconnection_policy.should_retry(attempts)
+
+                if delay:
+                    if delay is not True:
+                        await asyncio.sleep(delay)
```

Comment:
Reads funny and won't handle a delay of 0s.

I think for Python it might make more sense to have `float | None`, and a delay of 0 would be like the old `true`.